pkg/sql/sqlbase/index_encoding.go

```diff
@@ -164,6 +162,11 @@ func MakeKeyFromEncDatums(
 	alloc *DatumAlloc,
 ) (roachpb.Key, error) {
 	dirs := index.ColumnDirections
+	if !index.Unique {
+		// Extra columns are all ascending (the zero value).
+		extraDirs := make([]IndexDescriptor_Direction, len(index.ExtraColumnIDs))
+		dirs = append(dirs, extraDirs...)
```

Comment:
I'm sorry to keep harping on the allocations, but I really do think this is a big deal - an allocation per key build is not something we want to have to do. You should pass in the directions in the arguments. Or introduce a variant that grabs it from the index descriptor and passes it on if you don't want to change the callers.

Reply:
No need to apologize, not sure how I convinced myself this was OK. I converted it to an argument and did some rejiggering in related code to be able to pass it in easily.